File: src/test/java/cspfj/priorityqueues/SoftHeapTest.java
package cspfj.priorityqueues;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.Queue;
import java.util.Random;

import org.junit.Test;

public class SoftHeapTest {

    private static int id = 0;

    private static final Random RANDOM = new Random();

    private static final IdInteger[] INTS;

    static {
        INTS = new IdInteger[100000];
        for (int i = INTS.length; --i >= 0;) {
            INTS[i] = new IdInteger(RANDOM.nextInt(5000000));
        }
    }

    @Test
    public final void test() {
        final Queue<IdInteger> maximier = new SoftHeap<IdInteger>(
                new Key<IdInteger>() {

                    @Override
                    public double getKey(final IdInteger object) {
                        return object.value;
                    }

                });

        assertEquals(maximier.size(), 0);
        assertTrue(maximier.isEmpty());

        for (IdInteger i : INTS) {
            maximier.offer(i);
        }
        assertEquals(maximier.size(), INTS.length);

        int errors = 0;
        int last = maximier.peek().value;
        while (!maximier.isEmpty()) {
            final int current = maximier.poll().value;
            // System.out.println(current);
            // assertTrue(current + " should be >= " + last, current >= last);
            if (current < last) {
                errors++;
            }
            last = current;

        }
        System.out.println(errors + " errors (rate = " + (float) errors
                / INTS.length + ")");

    }

    private static class IdInteger implements Identified {

        private final int value;

        private final int id;

        public IdInteger(final int value) {
            this.value = value;
            this.id = SoftHeapTest.id++;
        }

        @Override
        public int getId() {
            return id;
        }

        public String toString() {
            return Integer.toString(value);
        }
    }
}

File: src/test/scala/cspfj/priorityqueues/SoftHeapTest.scala
package cspfj.priorityqueues;

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import java.util.Queue
import org.junit.Test
import cspfj.util.Loggable
import scala.util.Random

final class SoftHeapTest extends Loggable {

  val RANDOM = new Random()

  private val INTS = Stream.continually(IntNode(RANDOM.nextInt(5000000))).take(100000)

  @Test
  def test() {
    val maximier = new SoftHeap[IntNode](
      IntNode.key);

    assertEquals(maximier.size, 0);
    assertTrue(maximier.isEmpty);

    INTS.foreach(maximier.offer)

    assertEquals(maximier.size, INTS.length);

    var errors = 0;
    var last = maximier.poll().v;
    while (!maximier.isEmpty()) {
      val current = maximier.poll().v;
      // System.out.println(current);
      // assertTrue(current + " should be >= " + last, current >= last);
      if (current < last) {
        errors += 1;
      }
      last = current;

    }
    info(errors + " errors (rate = " + errors.toDouble / INTS.length + ")");

  }

}
